src/widget/container.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A convenience widget that combines common styling and positioning widgets.
+
+use crate::shell::kurbo::{Point, Rect, Size};
+use crate::shell::piet::{PaintBrush, RenderContext, StrokeStyle};
+use crate::widget::Padding;
+use crate::widget::SizedBox;
+use crate::{
+    BaseState, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget,
+};
+
+use std::marker::PhantomData;
+
+struct BorderState {
+    width: f64,
+    brush: PaintBrush,
+    style: StrokeStyle,
+}
+
+#[derive(Default)]
+struct ContainerStyle {
+    color: Option<PaintBrush>,
+    border: Option<BorderState>,
+}
+
+/// A convenience widget that combines common styling and positioning widgets.
+pub struct Container<T: Data> {
+    padding: f64,
+    style: ContainerStyle,
+    phantom: PhantomData<T>,
+}
+
+impl<T: Data + 'static> Container<T> {
+    pub fn new() -> Self {
+        Self {
+            padding: 0.0,
+            style: ContainerStyle::default(),
+            phantom: PhantomData::default(),
+        }
+    }
+
+    pub fn child(self, inner: impl Widget<T> + 'static) -> impl Widget<T> {
+        Padding::uniform(
+            self.border_padding(),
+            ContainerRaw::new(self.style, Padding::uniform(self.padding, inner)),
+        )
+    }
+
+    pub fn empty(self) -> impl Widget<T> {
+        self.child(SizedBox::empty())
+    }
+
+    pub fn color(mut self, brush: impl Into<PaintBrush>) -> Self {
+        self.style.color = Some(brush.into());
+        self
+    }
+
+    pub fn border(mut self, brush: impl Into<PaintBrush>, width: f64) -> Self {
```

Comment:
I think we need to either let the user set the stroke style or else not include it in the type.

Reply:
I will drop the stroke for now. Once we have a need it could be added.